File: ros2/temp/temp.cpp
#include <functional>
#include <memory>
#include <string>

#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/string.hpp"

using std::placeholders::_1;
using namespace std::chrono_literals;

class TempNode : public rclcpp::Node {
    rclcpp::Publisher<std_msgs::msg::String>::SharedPtr str_pub_;
    rclcpp::Subscription<std_msgs::msg::String>::SharedPtr str_sub_;
    rclcpp::TimerBase::SharedPtr timer_;

   public:
    TempNode()
        : Node("temp") {
        str_pub_ = this->create_publisher<std_msgs::msg::String>("pub", 10);
        str_sub_ = this->create_subscription<std_msgs::msg::String>(
            "sub", 10, std::bind(&TempNode ::str_cb, this, _1));
        timer_ = this->create_wall_timer(500ms, std::bind(&TempNode::timer_cb, this));
    }

   private:
    void timer_cb() {
        RCLCPP_INFO(this->get_logger(), "run");
    }
    void str_cb(const std_msgs::msg::String::SharedPtr msg) {
        auto message = std_msgs::msg::String();
        message.data = "I heard " + msg->data;
        str_pub_->publish(message);
    }
};

int main(int argc, char* argv[]) {
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<TempNode>());
    rclcpp::shutdown();
    return 0;
}

File: ros2/temp/temp.py
# -*- coding: utf-8 -*-
import rclpy
from rclpy.node import Node

from std_msgs.msg import String

class TempNode(Node):

    def __init__(self):
        super().__init__('temp')
        self.str_pub_ = self.create_publisher(String, 'tx', 10)
        self.str_sub_ = self.create_subscription(String,'rx',self.str_cb, 10)
        timer_period = 0.  # seconds
        self.timer = self.create_timer(timer_period, self.timer_cb)
        self.i = 0

    def timer_cb(self):
        msg = String()
        msg.data = 'Hello World: %d' % self.i
        self.str_pub_.publish(msg)
        self.get_logger().info('Publishing: "%s"' % msg.data)
        self.i += 1
    
    def str_cb(self, msg):
        self.get_logger().info('I heard: "%s"' % msg.data)

def main(args=None):
    rclpy.init(args=args)
    node = TempNode()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()